### PR TITLE
[feat] Named Parameter Groups in `LearningRateMonitor`

### LINKED ISSUE #7981: Let the `LearningRateMonitor` log the actual parameter group names, if they exist
## 🚀 Feature

As a `PytorchLightning` user, I would like to monitor parameter group learning rates that have been logged under the parameter group's given name.

### Motivation

My use case has _many_ parameter groups associated with different heads/tasks, each of which is partitioned within a unique (and static) parameter group. When using the `LearningRateMonitor` all these learning rates end up taking the form: `[pg1, pg2, ..., pgn]` ... making it hard/impossible to keep track of which `pg` maps to which task. See the attached TB screen shot below, to see what I mean.

### Pitch

Instead of `[pg1, pg2, ..., pgn]`, I want to take advantage of the existing parameter group naming functionality in `pytorch` like so:

```python
    class TestModel(BoringModel):

        def configure_optimizers(self):
            optimizer = torch.optim.SGD(
                [
                    {'params': [p for p in self.layer.parameters()], 'name': 'linear'}
                ],
                 lr=0.1
            )
            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
            return [optimizer], [lr_scheduler]
```

### Alternatives

Alternatively, I write down in my log book what heads `[pg1, pg2, ..., pgn]` map to. 

### Additional context

Screen Shot 2021-06-14 at 6 15 30 PM

### PR DESCRIPTION
**Pull request by AffineParameter**
## What does this PR do?

This PR allows the `LearningRateMonitor` callback to log the actual parameter group names, if they exist. It should fallback to prior behavior in the event that the parameter group names do not exist.

Fixes #7981

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/callbacks/test_lr_monitor.py
Comment (AffineParameter) [inline tests/callbacks/test_lr_monitor.py]: Any suggestions for how to stop playing pre-commit whack-a-mole?

Comment (AffineParameter) [reply tests/callbacks/test_lr_monitor.py]: Eh, looks like I got it.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_monitor.py:224]: This becomes an elif - else block, let's remove the indentation level

**Review by awaelchli — APPROVED** [review]

**Comment by AffineParameter** [comment]
Happy to update the CHANGELOG, provided I can get a milestone to target. This should tie a bow on things?

Thanks for the quick reviews!

**Comment by awaelchli** [comment]
@AffineParameter (love the name btw)  <3 
This is a feature so it will go into 1.4. Could you add the changelog entry to the 1.4 section then in CHANGELOG.md?

**Comment by AffineParameter** [comment]
PR Digression:

The contribution process (docs, workflow, reviews, etc.) have been really easy to work with! I am sure getting to this point hasn't been easy, but the effort and intentionality really shows. 🎩 's to all those responsible! :)

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (kaushikb11) [inline pytorch_lightning/callbacks/lr_monitor.py:231]:
```suggestion
        return set(n for n in names if names.count(n) > 1)
```

Comment (kaushikb11) [inline pytorch_lightning/callbacks/lr_monitor.py:231]:
```suggestion
        return set(n for n in names if names.count(n) > 1)
```

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:82]: nit: less use less whitespace for the docs

```suggestion
                [{
                    'params': [p for p in self.parameters()],
                    'name': 'my_parameter_group_name'
                }],
                lr=0.1
```

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:223]: nit: unnecessarsy else

```suggestion
        return name
```

File: tests/callbacks/test_lr_monitor.py
Comment (carmocca) [inline tests/callbacks/test_lr_monitor.py:300]: This will run 10% and 50% of the available batches respectively, do we need that many?

You can use `int` to set a specific number of batches

Same for the test below

File: tests/callbacks/test_lr_monitor.py
Comment (carmocca) [inline tests/callbacks/test_lr_monitor.py:307]: nit:

```suggestion
    assert list(lr_monitor.lrs) == ['lr-SGD/linear']
```

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:226]:
```suggestion
        names = [pg.get('name', f'pg{i}') for i, pg in enumerate(1, param_groups)]
```

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:225]: Maybe use a `collections.Counter` for efficiency instead of generating the list and then calling count for each element

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:260]: Could be one-line :D

```suggestion
            names.extend(self._add_suffix(name, param_groups, i) for i in range(len(param_groups)))
```

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [inline pytorch_lightning/callbacks/lr_monitor.py:212]: nit: do you see value in the full explicit name?

```suggestion
    def _add_suffix(self, name: str, param_groups: List[Dict], index: int, use_names: bool = True) -> str:
```

**Comment by AffineParameter** [comment]
This is the last time I draw attention to my PR's by complementing the maintainers... 🤣 

But seriously, I'll respond when I get the chance.

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py]: I explicitly avoided this pattern for perf. implications. Ideally, the `_duplicate_param_group_names` should be a validation check _only_ for the happy path, which is satisfied with the following check:

```python
 unique = set(names)        
 if len(names) == len(unique):
     return set()
 ```

This should be very fast as it will (optionally) get called every `on_train_batch_start`. The failure case will ultimately raise an exception, but I wanted the exception to provide actionable info on what ultimately caused the issue "the duplicate param-group names" in this case.

```python
else:
    return set(n for n in names if names.count(n) &gt; 1)
```

For this, I chose to trade off performance for readability as, mentioned above, an exception will ultimately be raised in this case, disallowing repeated calls.

In sum, I prefer the original implementation; but it's your codebase :)

Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py]: I am also open to being wrong because I missed something 😅

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py:222]: See: https://github.com/PyTorchLightning/pytorch-lightning/pull/7987#discussion_r652874708 for my thoughts on this.

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py:223]: I can't believe I have made it this far w/out learning about this before: https://docs.python.org/3.8/library/functions.html#enumerate

Thanks!

File: tests/callbacks/test_lr_monitor.py
Comment (AffineParameter) [reply tests/callbacks/test_lr_monitor.py:300]: Makes sense, these were clearly copy/paste UTs ;)

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py:222]: You can see my thoughts on this here: https://github.com/PyTorchLightning/pytorch-lightning/pull/7987#discussion_r652874708

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py:257]: I'm on the fence here. I think I prefer the readability++ of the intermediate variable name, but I don't feel strongly enough about it to oppose those who would prefer this suggestion.

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py:210]: My experience, I think, biases me to explicit names, especially for `kwargs` in particular. But, as I am a guest here I am happy to defer to the collective's preferences.

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (kaushikb11) [reply pytorch_lightning/callbacks/lr_monitor.py]: Haha, good enough reason! Go for it for the changes! 😄

Comment (AffineParameter) [reply pytorch_lightning/callbacks/lr_monitor.py]: I do appreciate the suggestion! Your suggestion def. lowers the surface area, which is an understandable priority for plugged-in maintainers!!

File: tests/callbacks/test_lr_monitor.py
Comment (AffineParameter) [reply tests/callbacks/test_lr_monitor.py:300]: Significant tests speedup! Excellent suggestion!

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (tchaton) [reply pytorch_lightning/callbacks/lr_monitor.py:261]: Hey @AffineParameter, we would prefer if you could use list comprehension there :)

**Comment by tchaton** [comment]
> PR Digression:
> 
> The contribution process (docs, workflow, reviews, etc.) have been really easy to work with! I am sure getting to this point hasn't been easy, but the effort and intentionality really shows. 🎩 's to all those responsible! :)

Hey @AffineParameter, we definitely appreciated your mindset and contribution !
We hope to see more contributions from you and hope to keep the contribution process at the same high standard.

Best,
T.C

File: pytorch_lightning/callbacks/lr_monitor.py
Comment (carmocca) [reply pytorch_lightning/callbacks/lr_monitor.py:222]: So my proposal was the following

```python
    def _duplicate_param_group_names(self, param_groups: List[Dict]) -> Set[str]:
        names = [pg.get('name', f'pg{i}') for i, pg in enumerate(param_groups, start=1)]
        counter = Counter(names)
        return set() if len(counter) == len(names) else {k for k, v in counter.items() if v > 1}
```

But I guess the Counter performance does not really matter since the else will execute only before erroring out

**Pull request merged**
